tests/unit/jobserver/test_forms.py

```diff
@@ -104,27 +104,42 @@ def test_jobrequestcreateform_with_bad_codelists(
     ],
 )
 def test_workspacecreateform_success(name, cleaned_name):
+    """
+    When the form is instantiated with valid data then:
+        * The form validates.
+        * The data ends up in the cleaned_data.
+        * The name field is cleaned by converting to lower case.
+    """
     project = ProjectFactory()
-    data = {
-        "name": name,
-        "repo": "http://example.com/derp/test-repo",
-        "branch": "test-branch",
-        "purpose": "test",
-    }
     repos_with_branches = [
         {
             "name": "test-repo",
             "url": "http://example.com/derp/test-repo",
             "branches": ["test-branch"],
-        }
+        },
+        {
+            "name": "test-repo2",
+            "url": "http://example.com/derp/test-repo2",
+            "branches": ["test-branch2"],
+        },
     ]
-    form = WorkspaceCreateForm(project, repos_with_branches, data)
+    data = {
+        "name": name,
+        "repo": "http://example.com/derp/test-repo2",
+        "branch": "test-branch2",
+        "purpose": "test purpose",
+    }
+    form = WorkspaceCreateForm(project, repos_with_branches, data=data)
```

Comment:
> These test cases were incorrectly passing `data` as a positional argument. A `Form` is initialised with POST data through the `data` _kwarg_.

This surprised me, Mike. Can you point to some documentation? It's not apparent from "[The Forms API](https://docs.djangoproject.com/en/5.1/ref/forms/api/)" page in the Django documentation that `MyForm(data)` is incorrect, but `MyForm(data=data)` is correct.

Reply:
Thanks, Iain, this is a good spot. I've indeed got that a bit wrong. The Django `Form` API allows either.

I could have said "our custom `WorkspaceCreateForm` only correctly handles POSTED `data` when supplied as a keyword-argument. This is because the class `__init__ ` code refers to `kwargs` directly." That would have been accurate. That's why I was finding my expanded success test was failing when `data` was passed by position.

But I'd gotten the implication of that the wrong way around -- the tests weren't broken, the form code design is (both before this PR, and with the changes as you initially reviewed them, as I continued the pattern of referring to `kwargs`.) It should access `self.data`, not subtly require the form to be instantiated with `data` passed by key-word to work correctly. It's not following the `Form` API properly. I'll fix.

This won't have been noticed so far because it's only used in one place in the prod code, and the test code was not robust enough to detect that the `data` it tried to supply was just being ignored in those tests.

Reply:
Committed a rebased fix. The [Use form.data, not kwargs. Validate more in tests.](https://github.com/opensafely-core/job-server/pull/4682/commits/07cbf7ec293cf23d9ec62615869e28c2ff7a54dc) commit is now a bit overloaded / non-atomic but it would be a bit of effort to unpick it. Let me know if you think I should.

Reply:
I need to rewrite the commit messages to reflect this. And the docstrings and comments can use some rework. But I think the form code and test code are reviewable.

I also think it's worth the tests explicitly checking the choice fields are constructed with the right options (would have caught this issue a different way if robust enough -- should have multiple repos and branches) and that `form.bound` is as we expect in each (which wouldn't have caught this issue but is worth sense-checking). What do you think?